File: App_Resources/Android/src/main/java/com/faceclaw/app/FaceclawSettings.java
package com.faceclaw.app;

import android.content.Context;
import android.util.AtomicFile;
import org.json.JSONObject;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import java.util.concurrent.CopyOnWriteArrayList;

/**
 * App settings store shared by every JS isolate (main thread and app
 * workers). JSONC encoding, typed values and migrations live in the shared
 * Kotlin SettingsDocument. This adapter serializes access and commits atomically.
 *
 * Change notifications: each isolate registers one listener from its own
 * thread. The registering thread's Looper is captured, and notifications are
 * posted through it so the JS callback always runs on the isolate's own
 * thread (calling into an isolate from a foreign thread is not allowed).
 * NativeScript worker threads run a message loop, so both the main thread
 * and workers have a Looper; a listener registered from a Looper-less thread
 * is accepted but never notified (it can still read fresh values on demand).
 */
public final class FaceclawSettings {
    private static final String TAG = "FaceclawSettings";
    private static final String PREFS_NAME = "faceclaw_settings";
    private static volatile FaceclawSettings instance;

    private final AtomicFile file;
    private SettingsDocument document;
    private final CopyOnWriteArrayList<ListenerEntry> listeners = new CopyOnWriteArrayList<>();

    private static final class ListenerEntry {
        final FaceclawSettingsListener listener;
        final Handler handler;

        ListenerEntry(FaceclawSettingsListener listener, Handler handler) {
            this.listener = listener;
            this.handler = handler;
        }
    }

    private FaceclawSettings(Context context) {
        this.file = new AtomicFile(configFile(context));
        try {
            // AtomicFile also recovers an interrupted write before we consider XML.
            if (file.getBaseFile().exists() || new File(file.getBaseFile() + ".bak").exists()) {
                String original = new String(file.readFully(), StandardCharsets.UTF_8);
                document = decode(original);
                // Persist a migration, but preserve comments on already-current files.
                if (new SettingsCodec().needsMigration(original)) persist(document);
            } else {
                JSONObject legacy = new JSONObject(context.getApplicationContext()
                        .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).getAll());
                // These keys were still stored through NativeScript's separate preferences.
                for (java.util.Map.Entry<String, ?> entry : context.getSharedPreferences("prefs.db", Context.MODE_PRIVATE).getAll().entrySet()) {
                    String key = entry.getKey();
                    if (key.startsWith("deviceAddress.") || key.startsWith("deviceIdentity.") || key.startsWith("onboarding.") || key.startsWith("evenhub:")) {
                        Object value = entry.getValue();
                        if (key.equals("deviceIdentity.pairedAtMs") && value instanceof Long) value = Double.longBitsToDouble((Long) value);
                        if (!legacy.has(key)) legacy.put(key, value);
                    }
                }
                document = decode("{\"schema\":1,\"settings\":" + legacy + "}");
                persist(document); // Never delete the legacy XML.
            }
        } catch (Exception error) {
            throw new IllegalStateException("Could not load settings; original config retained", error);
        }
    }

    /** Initialize (idempotent) and return the singleton. */
    public static FaceclawSettings getInstance(Context context) {
        if (instance == null) {
            synchronized (FaceclawSettings.class) {
                if (instance == null) {
                    instance = new FaceclawSettings(context);
                }
            }
        }
        return instance;
    }

    /** Return the singleton; the main isolate must have initialized it first. */
    public static FaceclawSettings getInstance() {
        FaceclawSettings result = instance;
        if (result == null) {
            throw new IllegalStateException("FaceclawSettings not initialized; call getInstance(context) first");
        }
        return result;
    }

    public static File configFile(Context context) {
        return new File(context.getFilesDir(), "faceclaw_settings.jsonc");
    }

    private static SettingsDocument decode(String text) throws IOException {
        SettingsDocument value = new SettingsCodec().decode(text);
        if (value == null) throw new IOException("Invalid or unsupported Faceclaw settings config");
        return value;
    }

    private void persist(SettingsDocument next) throws IOException {
        String encoded = next.encodeForStorage();
        if (encoded == null) throw new IOException("Settings exceed format limits");
        FileOutputStream stream = null;
        try {
            stream = file.startWrite();
            stream.write(encoded.getBytes(StandardCharsets.UTF_8));
            file.finishWrite(stream);
        } catch (IOException error) {
            file.failWrite(stream);
            throw error;
        }
        document = next;
    }

    public synchronized String exportConfig() { return document.encode(); }

    public synchronized void importConfig(String text) throws IOException {
        SettingsDocument next = decode(text);
        AtomicFile backup = new AtomicFile(new File(file.getBaseFile() + ".previous"));
        FileOutputStream stream = backup.startWrite();
        try { stream.write(file.readFully()); backup.finishWrite(stream); }
        catch (IOException error) { backup.failWrite(stream); throw error; }
        persist(next);
    }

    public synchronized String getString(String key, String defaultValue) {
        return document.getString(key, defaultValue);
    }

    public synchronized void setString(String key, String value) {
        if (document.containsString(key, value)) return;
        try { persist(document.replacingString(key, value)); }
        catch (IOException error) { throw new IllegalStateException("Could not save settings", error); }
        notifyChanged(key);
    }

    public synchronized boolean getBoolean(String key, boolean defaultValue) {
        return document.getBoolean(key, defaultValue);
    }

    public synchronized void setBoolean(String key, boolean value) {
        if (document.getBoolean(key, !value) == value) return;
        try { persist(document.replacingBoolean(key, value)); }
        catch (IOException error) { throw new IllegalStateException("Could not save settings", error); }
        notifyChanged(key);
    }

    public synchronized double getNumber(String key, double fallback) { return document.getNumber(key, fallback); }
    public synchronized void setNumber(String key, double value) {
        if (!Double.isFinite(value)) throw new IllegalArgumentException("Non-finite setting");
        try { persist(document.replacingNumber(key, value)); }
        catch (IOException error) { throw new IllegalStateException("Could not save settings", error); }
        notifyChanged(key);
    }
    public synchronized void remove(String key) {
        try { persist(document.removing(key)); }
        catch (IOException error) { throw new IllegalStateException("Could not save settings", error); }
        notifyChanged(key);
    }

    /**
     * Register a change listener. Must be called from the thread whose
     * isolate owns the listener; that thread's Looper is captured for
     * dispatch.
     */
    public void registerListener(FaceclawSettingsListener listener) {
        Looper looper = Looper.myLooper();
        if (looper == null) {
            Log.w(TAG, "settings listener registered from a Looper-less thread; it will never be notified");
        }
        listeners.add(new ListenerEntry(listener, looper != null ? new Handler(looper) : null));
    }

    public void unregisterListener(FaceclawSettingsListener listener) {
        for (ListenerEntry entry : listeners) {
            if (entry.listener == listener) {
                listeners.remove(entry);
            }
        }
    }

    private void notifyChanged(String key) {
        for (ListenerEntry entry : listeners) {
            if (entry.handler == null) continue;
            entry.handler.post(() -> {
                try {
                    entry.listener.onSettingChanged(key);
                } catch (Exception e) {
                    Log.w(TAG, "settings listener failed for key " + key, e);
                }
            });
        }
    }
}

File: App_Resources/Android/src/main/java/com/faceclaw/app/FaceclawSettings.kt
package com.faceclaw.app

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AtomicFile
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * App settings store shared by every JS isolate (main thread and app
 * workers). JSONC encoding, typed values and migrations live in the shared
 * Kotlin SettingsDocument. This adapter serializes access and commits atomically.
 *
 * Change notifications: each isolate registers one listener from its own
 * thread. The registering thread's Looper is captured, and notifications are
 * posted through it so the JS callback always runs on the isolate's own
 * thread (calling into an isolate from a foreign thread is not allowed).
 * NativeScript worker threads run a message loop, so both the main thread
 * and workers have a Looper; a listener registered from a Looper-less thread
 * is accepted but never notified (it can still read fresh values on demand).
 * The string/boolean reads/writes and the fan-out live in the shared SettingsChangeHub.
 */
class FaceclawSettings private constructor(context: Context) {
    private val file = AtomicFile(configFile(context))
    private lateinit var document: SettingsDocument
    // Only reached through the synchronized public methods below, which never pass null.
    private val hub = SettingsChangeHub(object : SettingsStorage {
        override fun getString(key: String?, defaultValue: String?): String? = document.getString(key!!, defaultValue!!)

        override fun putString(key: String?, value: String?) = save(document.replacingString(key!!, value!!))

        override fun getBoolean(key: String?, defaultValue: Boolean): Boolean = document.getBoolean(key!!, defaultValue)

        override fun putBoolean(key: String?, value: Boolean) = save(document.replacingBoolean(key!!, value))
    })

    init {
        try {
            // AtomicFile also recovers an interrupted write before we consider XML.
            if (file.baseFile.exists() || File("${file.baseFile}.bak").exists()) {
                val original = String(file.readFully(), Charsets.UTF_8)
                document = decode(original)
                // Persist a migration, but preserve comments on already-current files.
                if (SettingsCodec().needsMigration(original)) persist(document)
            } else {
                val legacy = JSONObject(context.applicationContext
                    .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).all)
                // These keys were still stored through NativeScript's separate preferences.
                for ((key, stored) in context.getSharedPreferences("prefs.db", Context.MODE_PRIVATE).all) {
                    if (key.startsWith("deviceAddress.") || key.startsWith("deviceIdentity.") || key.startsWith("onboarding.") || key.startsWith("evenhub:")) {
                        val value = if (key == "deviceIdentity.pairedAtMs" && stored is Long) Double.fromBits(stored) else stored
                        if (!legacy.has(key)) legacy.put(key, value)
                    }
                }
                document = decode("{\"schema\":1,\"settings\":$legacy}")
                persist(document) // Never delete the legacy XML.
            }
        } catch (error: Exception) {
            throw IllegalStateException("Could not load settings; original config retained", error)
        }
    }

    companion object {
        private const val PREFS_NAME = "faceclaw_settings"
        @Volatile
        private var instance: FaceclawSettings? = null

        /** Initialize (idempotent) and return the singleton. */
        @JvmStatic
        fun getInstance(context: Context): FaceclawSettings {
            if (instance == null) {
                synchronized(FaceclawSettings::class.java) {
                    if (instance == null) {
                        instance = FaceclawSettings(context)
                    }
                }
            }
            return instance!!
        }

        /** Return the singleton; the main isolate must have initialized it first. */
        @JvmStatic
        fun getInstance(): FaceclawSettings {
            val result = instance
                ?: throw IllegalStateException("FaceclawSettings not initialized; call getInstance(context) first")
            return result
        }

        @JvmStatic
        fun configFile(context: Context): File = File(context.filesDir, "faceclaw_settings.jsonc")

        @Throws(IOException::class)
        private fun decode(text: String): SettingsDocument =
            SettingsCodec().decode(text) ?: throw IOException("Invalid or unsupported Faceclaw settings config")
    }

    @Throws(IOException::class)
    private fun persist(next: SettingsDocument) {
        val encoded = next.encodeForStorage() ?: throw IOException("Settings exceed format limits")
        var stream: FileOutputStream? = null
        try {
            stream = file.startWrite()
            stream.write(encoded.toByteArray(Charsets.UTF_8))
            file.finishWrite(stream)
        } catch (error: IOException) {
            file.failWrite(stream)
            throw error
        }
        document = next
    }

    private fun save(next: SettingsDocument) {
        try {
            persist(next)
        } catch (error: IOException) {
            throw IllegalStateException("Could not save settings", error)
        }
    }

    @Synchronized
    fun exportConfig(): String = document.encode()

    @Synchronized
    @Throws(IOException::class)
    fun importConfig(text: String) {
        val next = decode(text)
        val backup = AtomicFile(File("${file.baseFile}.previous"))
        val stream = backup.startWrite()
        try {
            stream.write(file.readFully())
            backup.finishWrite(stream)
        } catch (error: IOException) {
            backup.failWrite(stream)
            throw error
        }
        persist(next)
    }

    @Synchronized
    fun getString(key: String, defaultValue: String): String? = hub.getString(key, defaultValue)

    @Synchronized
    fun setString(key: String, value: String) {
        if (!document.containsString(key, value)) hub.setString(key, value)
    }

    @Synchronized
    fun getBoolean(key: String, defaultValue: Boolean): Boolean = hub.getBoolean(key, defaultValue)

    @Synchronized
    fun setBoolean(key: String, value: Boolean) {
        if (document.getBoolean(key, !value) != value) hub.setBoolean(key, value)
    }

    @Synchronized
    fun getNumber(key: String, fallback: Double): Double = document.getNumber(key, fallback)

    @Synchronized
    fun setNumber(key: String, value: Double) {
        require(value.isFinite()) { "Non-finite setting" }
        save(document.replacingNumber(key, value))
        hub.notifyChanged(key)
    }

    @Synchronized
    fun remove(key: String) {
        save(document.removing(key))
        hub.notifyChanged(key)
    }

    /**
     * Register a change listener. Must be called from the thread whose
     * isolate owns the listener; that thread's Looper is captured for
     * dispatch.
     */
    fun registerListener(listener: FaceclawSettingsListener) {
        val looper = Looper.myLooper()
        val handler = if (looper != null) Handler(looper) else null
        hub.register(listener, if (handler != null) SettingsDispatcher { action -> handler.post(action) } else null)
    }

    fun unregisterListener(listener: FaceclawSettingsListener?) = hub.unregister(listener)
}
